service/src/main/java/org/cbioportal/service/impl/CoExpressionServiceImpl.java

```diff
@@ -28,54 +39,80 @@
+        GeneMolecularAlteration queryMolecularDataList = molecularDataMap.remove(queryEntrezGeneId);
 
         Map<Integer, List<Gene>> genes = geneService.fetchGenes(molecularDataMap.keySet().stream()
             .map(String::valueOf).collect(Collectors.toList()), "ENTREZ_GENE_ID", "SUMMARY").stream()
             .collect(Collectors.groupingBy(Gene::getEntrezGeneId));
         
         List<CoExpression> coExpressionList = new ArrayList<>();
-
         if (queryMolecularDataList == null) {
             return coExpressionList;
         }
 
-        List<String> queryValues = queryMolecularDataList.stream().map(g -> g.getValue()).collect(Collectors.toList());
+        String commaSeparatedSampleIdsOfMolecularProfile = molecularDataRepository
+            .getCommaSeparatedSampleIdsOfMolecularProfile(molecularProfileId);
+        List<Integer> internalSampleIds = Arrays.stream(commaSeparatedSampleIdsOfMolecularProfile.split(","))
+            .mapToInt(Integer::parseInt).boxed().collect(Collectors.toList());
+        Map<Integer, Integer> internalSampleIdsMap = new HashMap<>();
+        for (int lc = 0; lc < internalSampleIds.size(); lc++) {
+            internalSampleIdsMap.put(internalSampleIds.get(lc), lc);
+        }
+
+        MolecularProfile molecularProfile = molecularProfileService.getMolecularProfile(molecularProfileId);
+        List<String> studyIds = new ArrayList<>();
+        sampleIds.forEach(s -> studyIds.add(molecularProfile.getCancerStudyIdentifier()));
+        List<Sample> samples = sampleService.fetchSamples(studyIds, sampleIds, "ID");
+
+        Set<Integer> excludedIndexes = new HashSet<>();
+        for (Sample sample : samples) {
+            Integer indexOfSampleId = internalSampleIdsMap.get(sample.getInternalId());
+            if (indexOfSampleId == null) {
+                excludedIndexes.add(indexOfSampleId);
+            }
+        }
+
+        List<String> queryValues = Arrays.asList(queryMolecularDataList.getSplitValues());
         for (Integer entrezGeneId : molecularDataMap.keySet()) {
             
-            List<String> values = molecularDataMap.get(entrezGeneId).stream().map(g -> g.getValue())
-                .collect(Collectors.toList());
+            List<String> values = new ArrayList<>(Arrays.asList(molecularDataMap.get(entrezGeneId).getSplitValues()));
             List<String> queryValuesCopy = new ArrayList<>(queryValues);
 
             List<Integer> valuesToRemove = new ArrayList<>();
             for (int i = 0; i < queryValuesCopy.size(); i++) {
-                if (!NumberUtils.isNumber(queryValuesCopy.get(i)) || !NumberUtils.isNumber(values.get(i))) {
+                if (!NumberUtils.isNumber(queryValuesCopy.get(i)) || !NumberUtils.isNumber(values.get(i)) 
+                    || excludedIndexes.contains(i)) {
                     valuesToRemove.add(i);
                 }
             }
```

Comment:
I can't add a comment below but, we can do one further optimization. Below, instead of what you have for arrays, let it be like this:

```
double[][] arrays = new double[2][valuesNumber.length];
arrays[0] = queryValuesNumber;
arrays[1] = valuesNumber;
...
SpearmansCorrelation spearmansCorrelation = new SpearmansCorrelation((new Array2DRowRealMatrix(arrays, false)).transpose())
```

It seems wasteful to duplicate that data, right?

Reply:
@adamabeshouse I don't know much about matrix transformation but it sounds good to me, so I just updated the PR.